src/otk/ui.py

```diff
@@ -0,0 +1,69 @@
+import logging
+import sys
+import time
+import threading
+
+from typing import Optional, Type, Any
+from types import TracebackType
+
+from . import __version__
+
+
+class _Spinner:
+    """Render a spinner with an optional prompt to stderr, if stderr isn't a
+    tty this only prints the prompt."""
+
+    active: bool
+    thread: threading.Thread
+
+    def __init__(self, prompt: str) -> None:
+        self.prompt = prompt
+
+    def _loop(self):
+        sys.stderr.write(self.prompt)
+        sys.stderr.flush()
+
+        # No spinner business if we aren't printing to tty's.
+        if sys.stderr.isatty():
+            while self.active:
```

Comment:
(super nitpick) I wonder if these two lines should be flipped around (and the wait needs to be there of course). As it is now the code on a non-terminal will write "running foo" and then immediately "\n". Flipping around would mean it still waits without output and only when finished writes the \n (i.e. _loop does not really loop but exit right away).